Add https-certificate blocks to Cobalt Strike C2 profiles

jsf9k opened the pull request:
## 🗣 Description ##

This pull request makes the necessary changes to ensure that appropriate `https-certificate` configuration blocks are appended to the Amazon and OCSP Cobalt Strike C2 profiles from [rsmudge/Malleable-C2-Profiles](https://github.com/rsmudge/Malleable-C2-Profiles).

I also did some cleanup in commits 497126f, 6687dd4, and 7bc8218 to get rid of some needlessly duplicated resources.

## 💭 Motivation and context ##

Resolves #116.  See also cisagov/teamserver-packer#32.

## 🧪 Testing ##

I deployed these changes to our staging COOL environment and verified that they work as expected.

## ✅ Checklist ##

* [x] This PR has an informative and human-readable title.
* [x] Changes are limited to a single goal - _eschew scope creep!_
* [x] _All_ future TODOs are captured in issues, which are referenced in code comments.
* [x] All relevant type-of-change labels have been added.
* [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
* [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
* [x] All relevant repo and/or project documentation has been updated to reflect the changes in this PR.
* [x] All new and existing tests pass.

jsf9k committed:
Give teamserver IAM permissions to pull its certificate from a bucket

This necessitates the addition of a cisagov/cert-read-role-tf-module.

jsf9k committed:
Give teamserver its own cloud init configuration

This cloud-init configuration allows for mounting the EFS share and
installing the appropriate certificates.

jsf9k committed:
Remove unnecessary executable permissions

jsf9k committed:
Generalize cert installation script

The certificate installation certificate script now accepts as inputs
locations for the full certificate chain and private key PEM files.
This is done to make the script more generic, so that it can be used
for both the guacamole and teamserver instances.

jsf9k committed:
Add SG rules so teamservers can access the S3 and STS VPC endpoints

jsf9k committed:
Use only the domain to reference wildcard certs

It turns out that certbot stores wildcard certs in a directory with
the name of the domain, instead of pre-pending an asterisk.

jsf9k committed:
Add new cloud-init script to massage CS profiles

jsf9k committed:
Specify the correct location of the CS C2 profiles

dav3r commented on cloud-init/add-https-certificate-block-to-cs-profiles.tpl.sh:
```diff
@@ -0,0 +1,42 @@
+CERT_BLOCK
+
+cat /tmp/cert-block.txt >> "${c2_profile_location}/amazon.profile"
+cat /tmp/cert-block.txt >> "${c2_profile_location}/ocsp.profile"
```
Not sure if you are keeping it around for testing right now, but don't forget to `rm /tmp/cert-block.txt` at the end of this script.

jsf9k replied:
See commit 2b63d1c.

jsf9k committed:
Force order-dependent scripts to run in the correct order

jsf9k committed:
Remove unnecessary merge_type line

User scripts do not require such a thing.

jsf9k committed:
Alphabetize dictionary keys

jsf9k committed:
Delete temp file

Co-authored-by: Dave Redmin <david.redmin@beta.dhs.gov>

jsf9k committed:
Add an input to the cert installation script

This Boolean input determines whether or not the destination
directories for the full chain and private key PEMs are created.  In
the case of the Guacamole instance it makes sense to do this, but
normally you would want the script to error out if the destination
directory does not exist.

jsf9k committed:
Give assume role policy document data resource a more specific name

jsf9k committed:
Add an assume role policy that allows assumption by the AWS EC2 service

This is done to DRY out a piece of duplicated code.

jsf9k committed:
Create a single EFS mount policy and reuse it everywhere

Also use the EC2 service assume role policy document everywhere.

dav3r commented on teamserver_cloud_init.tf:
```diff
@@ -0,0 +1,78 @@
+  # take place in a certain order, so we prepend numbers to the script
+  # names to force that to that happen.
+  #
+  # Here is there the user scripts are called by cloud-init:
```
Typo:
```suggestion
  # Here is where the user scripts are called by cloud-init:
```

jsf9k committed:
Update README.md via terraform-docs

jsf9k committed:
Fix typo

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

mcdonnnj commented on cloud-init/add-https-certificate-block-to-cs-profiles.tpl.sh:
```diff
@@ -0,0 +1,45 @@
+# Convert the full chain and private key PEMs into a single P12 file.
+#
+# This is a Terraform template file, and the domain, full_chain_pem,
+# and priv_key_pem variables are passed in via the templatefile().
```
Alternatively something like `are passed in via the templatefile() call.`.

```suggestion
# and priv_key_pem variables are passed in via templatefile().
```

mcdonnnj commented on cloud-init/add-https-certificate-block-to-cs-profiles.tpl.sh:
```diff
@@ -0,0 +1,45 @@
+# Create the Java keystore from the P12 file.
+#
+# This is a Terraform template file, and the c2_profile_location and
+# domain variables are passed in via the templatefile().
```
Same train of thought.

```suggestion
# domain variables are passed in via templatefile().
```

mcdonnnj commented on ec2_service_assume_role_policy.tf:
```diff
@@ -0,0 +1,15 @@
+
+data "aws_iam_policy_document" "ec2_service_assume_role_doc" {
+  statement {
+    actions = ["sts:AssumeRole", ]
```
```suggestion
    actions = [
      "sts:AssumeRole",
      "sts:TagSession",
    ]
```

dav3r commented on cloud-init/install-certificates.py:
```diff
@@ -61,15 +64,19 @@
+# there.
+if CREATE_DEST_DIRS:
+    os.makedirs(os.path.dirname(FULL_CHAIN_PEM_DEST), exist_ok=True)
+    os.makedirs(os.path.dirname(PRIV_KEY_PEM_DEST), exist_ok=True)
```
Nice improvement and generalization. 👍

jsf9k replied:
Thank you, sir.  It was the least ugly way I could think of to support both cases.

dav3r commented on cloud-init/add-https-certificate-block-to-cs-profiles.tpl.sh:
```diff
@@ -0,0 +1,45 @@
+CERT_BLOCK
+
+cat /tmp/cert-block.txt >> "${c2_profile_location}/amazon.profile"
+cat /tmp/cert-block.txt >> "${c2_profile_location}/ocsp.profile"
```
I was thinking it would be cool to make the list of Cobalt Strike profiles to be modified an input to this script, which should include adding some error-handling (e.g. if it was fed a profile name that doesn't exist).  I don't think it's really necessary now- just wanted to mention it as a potential future improvement.

jsf9k replied:
I had a similar thought.  I'll create an issue and add a link.

jsf9k replied:
See #121 and commit b365cd5.

dav3r replied:
Solid - thanks for writing that up!

dav3r commented on teamserver_cloud_init.tf:
```diff
@@ -0,0 +1,78 @@
+  # https://github.com/canonical/cloud-init/blob/master/cloudinit/config/cc_scripts_user.py#L45
+  #
+  # And here is where you can see how cloud-init sorts the scripts:
+  # https://github.com/canonical/cloud-init/blob/master/cloudinit/subp.py#L373
```
What a great comment - I should steal it for my own use!

dav3r commented on teamserver_iam.tf:
```diff
@@ -1,5 +1,20 @@
+
 # Create the IAM instance profile for the Teamserver EC2 server
-# instances
+# instance
```
This instance profile can be used by any number of Teamserver instances- guessing that's why we had it pluralized before.  Maybe `instance(s)` would make that clearer?

jsf9k replied:
Not sure it makes a difference.  I only changed it because right now only one is supported (unless you assign them all the same domain).  If you really want me to change it back I don't mind doing it.

dav3r replied:
Doesn't make a huge difference to me- I'm fine leaving it as is.

dav3r reviewed (approved):
This is great work - way to go above and beyond with some excellent cleanup of repeated code! 🥇

jsf9k committed:
Add another shellcheck-disabling line

shellcheck doesn't actually require this line, but it's good to
document the origin of c2_profile_location.  It's also good to include
this disable in case the one above were to get deleted, since that
would then make this line necessary.

jsf9k committed:
Add comments referencing issue #121 where applicable

felddy commented on cloud-init/install-certificates.py:
```diff
@@ -61,15 +64,19 @@
+# Ensure destination directories exist before we put the cert files
+# there.
+if CREATE_DEST_DIRS:
+    os.makedirs(os.path.dirname(FULL_CHAIN_PEM_DEST), exist_ok=True)
+    os.makedirs(os.path.dirname(PRIV_KEY_PEM_DEST), exist_ok=True)
```
Consider using `pathlib` for these calls: 
https://docs.python.org/3/library/pathlib.html#pathlib.Path.mkdir

jsf9k replied:
See commit 7be3aba.

felddy replied:
I've been swimming in raw sewage.  I love it! ❤️

felddy commented on teamserver_cloud_init.tf:
```diff
@@ -0,0 +1,78 @@
+        server_fqdn = var.email_sending_domain
+    })
+    content_type = "text/x-shellscript"
+    filename     = "01-install-certificates.py"
```
I bring a bit of my BASIC skillz in here... Would it make sense to use `10-` and `20-` instead of `01-` and `02-`  to allow for future code to exist in between these scripts?  or  `010-`, `020-`

jsf9k replied:
I think we can just renumber the filenames when that happens.  The only necessary changes are in this file.

felddy reviewed (approved):
Looks gud!   
![iu](https://user-images.githubusercontent.com/3229435/119887906-b99f5980-bf02-11eb-9422-f18d7cc695ef.jpeg)

mcdonnnj commented on teamserver_cloud_init.tf:
```diff
@@ -0,0 +1,78 @@
+        cert_bucket_name    = var.cert_bucket_name
+        cert_read_role_arn  = module.teamserver_certreadrole.role.arn
+        create_dest_dirs    = false
+        full_chain_pem_dest = "/tmp/fullchain.pem"
+        priv_key_pem_dest   = "/tmp/privkey.pem"
```
Should these be `local`s since they are also reused below?

jsf9k replied:
See commit 387110c.

mcdonnnj commented on teamserver_ec2.tf:
```diff
@@ -28,8 +28,21 @@ resource "aws_instance" "teamserver" {
+  # security group rules that allow STS endpoint access from the
+  # Teamserver, as well as the endpoints themselves.  Note that there
+  # is no security group rule for S3 because it's a _gateway_
+  # endpoint, while STS is a _interface_ endpoints.
```
```suggestion
  # endpoint, while STS is an _interface_ endpoints.
```

jsf9k replied:
See commit 6dbdd87.

mcdonnnj replied:
Ah thanks for catching the pluralization.

mcdonnnj commented on teamserver_ec2.tf:
```diff
@@ -55,7 +68,7 @@ resource "aws_instance" "teamserver" {
   }
   # We can use the same cloud-init code as the Kali instances, since
   # all it does is set up /etc/fstab to mount the EFS file share.
```
Does this comment still apply?

jsf9k replied:
See commit ffded2a.

mcdonnnj commented on teamserver_iam.tf:
```diff
@@ -35,13 +56,27 @@ resource "aws_iam_role_policy_attachment" "ssm_agent_policy_attachment_teamserve
+# the teamserver certificates from an S3 bucket.
+data "aws_iam_policy_document" "teamserver_assume_delegated_role_policy_doc" {
+  statement {
+    actions = ["sts:AssumeRole"]
```
```suggestion
    actions = [
      "sts:AssumeRole",
      "sts:TagSession",
    ]
```

mcdonnnj reviewed (approved):
This is some great work! I did have some comment questions and a few suggestions for your consideration.

jsf9k committed:
Improve English

Co-authored-by: Nick M. <50747025+mcdonnnj@users.noreply.github.com>

jsf9k committed:
Add TagSession permission

Thanks to @mcdonnnj, who keeps catching these omissions on my part!

Co-authored-by: Nick M. <50747025+mcdonnnj@users.noreply.github.com>

jsf9k committed:
Remove comment that is no longer truthful

Co-authored-by: Nick M <50747025+mcdonnnj@users.noreply.github.com>

jsf9k committed:
Use correct English

Co-authored-by: Nick M <50747025+mcdonnnj@users.noreply.github.com>

jsf9k committed:
Define a few local variables, so we can define values in one place

Thanks to @mcdonnnj for pointing this out.

Co-authored-by: Nick M <50747025+mcdonnnj@users.noreply.github.com>

jsf9k committed:
Prefer pathlib.Path to os,makedirs and os.path.dirname

Co-authored-by: Mark Feldhousen <felddy@users.noreply.github.com>

felddy reviewed (approved):
![tenor-3](https://user-images.githubusercontent.com/3229435/120001102-73e59e00-bfa1-11eb-8ebc-047993bc457f.gif)

mcdonnnj commented on cloud-init/add-https-certificate-block-to-cs-profiles.tpl.sh:
```diff
@@ -0,0 +1,59 @@
+
+# Create the Java keystore from the P12 file.
+#
+# TODO: issue #121 suggests improving this code by allowing the user
+# to specify the location of the Java keystore.
```
I would like where the issue is included at a minimum just so it's specifically mentioned. These issues aren't always in the repos they're mentioned so I think it's better to be explicit.

```suggestion
# TODO: issue cisagov/cool-assessment-terraform#121 suggests improving
# this code by allowing the user to specify the location of the Java
# keystore.
```

mcdonnnj commented on cloud-init/add-https-certificate-block-to-cs-profiles.tpl.sh:
```diff
@@ -0,0 +1,59 @@
+# Append the https-certificate blocks to the Cobalt Strike C2
+# profiles.
+#
+# TODO: issue #121 suggests improving this code by allowing the user
+# to specify the Cobalt Strike C2 profiles to which an
+# https-certificate block should be added.
```
Same rationale as above.

```suggestion
# TODO: issue cisagov/cool-assessment-terraform#121 suggests improving
# this code by allowing the user to specify the Cobalt Strike C2 profiles
# to which an https-certificate block should be added.
```

mcdonnnj reviewed (approved):
Thanks for continuing to refine this 💪. I had two more suggestions I missed the first pass, but otherwise it's looking nice.

jsf9k committed:
Use a "fuller" issue reference to provide more context

Co-authored-by: Nick M. <50747025+mcdonnnj@users.noreply.github.com>
